File: src/main/java/cspom/compiler/patterns/DiffGe.java
package cspom.compiler.patterns;

import java.util.List;
import java.util.NoSuchElementException;

import com.google.common.collect.Iterables;

import cspom.CSPOM;
import cspom.constraint.AbstractConstraint;
import cspom.constraint.CSPOMConstraint;
import cspom.constraint.FunctionalConstraint;
import cspom.constraint.GeneralConstraint;
import cspom.variable.CSPOMVariable;

/**
 * If constraint is the sub() constraint, converts a=sub(y,z), x=abs(a) to
 * x=absdiff(y,z). No other constraint may imply the auxiliary constraint a.
 */
public final class DiffGe implements ConstraintCompiler {
    private final CSPOM problem;

    public DiffGe(final CSPOM problem) {
        this.problem = problem;
    }

    @Override
    public void compile(final CSPOMConstraint constraint) {
        if (!("sub".equals(constraint.getDescription()) && constraint instanceof FunctionalConstraint)) {
            return;
        }
        final FunctionalConstraint subConstraint = (FunctionalConstraint) constraint;
        final CSPOMVariable result = subConstraint.getResultVariable();
        if (!result.isAuxiliary() || result.getConstraints().size() != 2) {
            return;
        }
        final CSPOMConstraint geConstraint = geConstraint(result);
        if (geConstraint == null) {
            return;
        }
        problem.removeConstraint(subConstraint);
        problem.removeConstraint(geConstraint);

        if (geConstraint instanceof FunctionalConstraint) {
            final CSPOMVariable[] scope = new CSPOMVariable[] {
                    subConstraint.getVariable(1), subConstraint.getVariable(2),
                    geConstraint.getVariable(2) };
            problem.addConstraint(new FunctionalConstraint(geConstraint
                    .getVariable(0), "diffGe", null, scope));
        } else {
            final CSPOMVariable[] scope = new CSPOMVariable[] {
                    subConstraint.getVariable(1), subConstraint.getVariable(2),
                    geConstraint.getVariable(1) };
            problem.addConstraint(new GeneralConstraint("diffGe", null, scope));
        }
    }

    private static CSPOMConstraint geConstraint(final CSPOMVariable variable) {
        final CSPOMConstraint geConstraint;
        try {
            geConstraint = Iterables.find(variable.getConstraints(),
                    AbstractConstraint.matchesDescription("ge"));
        } catch (NoSuchElementException e) {
            return null;
        }
        final List<CSPOMVariable> scope;
        if (geConstraint instanceof FunctionalConstraint) {
            scope = ((FunctionalConstraint) geConstraint).getArguments();
        } else {
            scope = geConstraint.getScope();
        }
        if (scope.size() == 2 && scope.get(0) == variable) {
            return geConstraint;
        }
        return null;
    }

}

File: src/main/scala/cspom/compiler/patterns/DiffGe.scala
package cspom.compiler.patterns

import cspom.constraint.FunctionalConstraint
import cspom.constraint.CSPOMConstraint
import cspom.CSPOM
import cspom.constraint.GeneralConstraint

/**
 * Transforms x = sub(y, z), [t =] ge(x, k) into [t =] diffGe(y, z, k)
 */
final class DiffGe(val problem: CSPOM) extends ConstraintCompiler {

  override def compileFunctional(subConstraint: FunctionalConstraint) {
    if (subConstraint.description == "sub" &&
      subConstraint.result.auxiliary &&
      subConstraint.result.constraints.size == 2) {

      for (
        geConstraint <- subConstraint.result.constraints.find { c: CSPOMConstraint =>
          c.description == "ge" && {
            val scope = c match {
              case fGe: FunctionalConstraint => fGe.arguments
              case gGe: GeneralConstraint => gGe.scope
            }
            scope.size == 2 && scope(0) == subConstraint.result
          }
        }
      ) {

        problem.removeConstraint(subConstraint);
        problem.removeConstraint(geConstraint);
        problem.removeVariable(subConstraint.result)

        geConstraint match {
          case fc: FunctionalConstraint =>
            problem.addConstraint(new FunctionalConstraint(
              fc.result,
              "diffGe",
              subConstraint.arguments :+ fc.arguments(1): _*))
          case _ =>
            problem.addConstraint(new GeneralConstraint("diffGe",
              subConstraint.arguments :+ geConstraint.scope(1): _*))

        }
      }
    }

  }

}
